tomopy/prep/phase.py

```diff
@@ -119,36 +129,47 @@ def retrieve_phase(
     # Filter in Fourier space.
     phase_filter = np.fft.fftshift(
         _paganin_filter_factor(energy, dist, alpha, w2))
+    mx = phase_filter.max()
 
-    prj = val * np.ones((dy + 2 * py, dz + 2 * pz), dtype='float32')
-    arr = mproc.distribute_jobs(
-        tomo,
-        func=_retrieve_phase,
-        args=(phase_filter, py, pz, prj, pad),
-        axis=0,
-        ncore=ncore,
-        nchunk=nchunk)
-    return arr
+    ncore, nchunk = mproc.get_ncore_nchunk(dx, ncore, nchunk)
 
+    chnks = np.round(np.linspace(0, dx, ncore+1)).astype(np.int)
 
-def _retrieve_phase(tomo, phase_filter, px, py, prj, pad):
-    dx, dy, dz = tomo.shape
-    num_jobs = tomo.shape[0]
-    for m in range(num_jobs):
-        prj[px:dy + px, py:dz + py] = tomo[m]
-        prj[:px] = prj[px]
-        prj[-px:] = prj[-px-1]
-        prj[:,:py] = prj[:,py][:,np.newaxis]
-        prj[:,-py:] = prj[:,-py-1][:,np.newaxis]
-        fproj = pyfftw.interfaces.numpy_fft.fft2(
-                prj, planner_effort=_plan_effort(num_jobs))
-        filtproj = np.multiply(phase_filter, fproj)
-        proj = np.real(pyfftw.interfaces.numpy_fft.ifft2(
-            filtproj, planner_effort=_plan_effort(num_jobs))
-            ) / phase_filter.max()
-        if pad:
-            proj = proj[px:dy + px, py:dz + py]
-        tomo[m] = proj
+    prj = pyfftw.n_byte_align_empty((dx, dy + 2 * py, dz + 2 * pz), pyfftw.simd_alignment, dtype='complex64')
+    val = np.complex64(val)
+    e = cf.ThreadPoolExecutor(ncore)
+
+    def setv(a, b, c, l, r, u, d):
```

Comment:
Instead of using concurrent futures here, have you considered using numba vectorize with a parallel target?
http://numba.pydata.org/numba-doc/latest/user/vectorize.html